File: jysmol_typescript/src/parser.ts
export type JysmolType = JysmolObject | JysmolArray | JysmolPrimitive
export type JysmolPrimitive = number | string | null | boolean
export type JysmolObject = {[key: string]: JysmolType}
export type JysmolArray = JysmolType[]

export class JysmolParser {
    private position: number
    private ch!: string

    constructor(private input: string) {
        this.position = -1

        this.advance()
    }

    static parse(input: string): JysmolType {
        const p = new JysmolParser(input)

        return p.parseValue()
    }

    private parseValue(): JysmolType {
        this.skipWhitespace()

        switch (this.ch) {
            case '"':
                return this.parseString()
            case '{':
                return this.parseObject()
            case '[':
                return this.parseArray()
        }

        if (isDigit(this.ch)) return this.parseNumber()
        if (this.ch === '-') {
            this.advance()
            return -this.parseNumber()
        }

        if (isAlpha(this.ch)) return this.parseKeywordValue()

        throw new Error(`unexpected token: '${this.ch}', at position ${this.position}`)
    }

    private parseKeywordValue(): null | boolean {
        const start = this.position
        let end = this.position

        while (isAlpha(this.ch) && this.position < this.input.length) {
            end++
            this.advance()
        }

        const keyword = this.input.slice(start, end)
        switch (keyword) {
            case 'null': return null
            case 'true': return true
            case 'false': return false
        }

        throw new Error(`not allowed keyword: ${keyword}`)
    }

    private parseObject(): JysmolObject {
        const obj: JysmolObject = {}
        this.advance()
        this.skipWhitespace()

        while(this.ch !== '}' && this.position < this.input.length) {
            const key = this.parseString()
            this.skipWhitespace()
            this.eat(':')
            const value = this.parseValue()

            obj[key] = value

            this.skipWhitespace()
            this.eat(',')
            this.skipWhitespace()
        }

        this.skipWhitespace()

        this.eat('}')

        return obj
    }

    private parseArray(): JysmolArray {
        const arr: JysmolArray = []
        this.advance()
        this.skipWhitespace()

        while(this.ch !== ']' && this.position < this.input.length) {
            const value = this.parseValue()
            arr.push(value)

            this.skipWhitespace()
            this.eat(',')
            this.skipWhitespace()
        }

        this.eat(']')

        return arr
    }

    private parseNumber(): number {
        let numberComponents = this.parseIntegerLiteral()

        if (this.ch === '.' ) {
            this.advance()
            numberComponents += '.'
            const newIntegerLiteral = this.parseIntegerLiteral()
            if (!newIntegerLiteral) throw new Error(`. without digit`)
            numberComponents += newIntegerLiteral
        }

        return Number(numberComponents)
    }

    private parseIntegerLiteral(): string {
        const start = this.position
        let end = this.position

        while(isDigit(this.ch) && this.position < this.input.length) {
            end++
            this.advance()
        }

        return this.input.slice(start, end)
    }

    private parseString(): string {
        this.advance()
        const start = this.position
        let end = this.position

        while(this.ch !== '"' && this.position < this.input.length) {
            end++
            this.advance()
        }

        this.eat('"')

        return this.input.slice(start, end)
    }

    private eat(ch: string): void {
        if (this.ch !== ch) throw new Error(
            `expected: '${ch}', got: '${this.ch}', at position ${this.position}`
        )
        this.advance()
    }

    private advance(): void {
        this.position++
        if (this.position < this.input.length) {
            this.ch = this.input[this.position]
        } else {
            this.ch = '\0'
        }
    }

    private skipWhitespace(): void {
        while (isWhitespace(this.ch)) {
            this.advance()
        }
    }
}

function isWhitespace(ch: string): boolean {
    return ch === ' ' || ch === '\n' || ch === '\t' || ch === '\r'
}

function isDigit(ch: string): boolean {
    return ch >= '0' && ch <= '9'
}

function isAlpha(ch: string): boolean {
    return ch >= 'a' && ch <= 'z' || ch >= 'A' && ch <= 'Z'
}

File: jysmol_go/jysmol/parser.go
package jysmol

import (
	"fmt"
	"strconv"
)

type JysmolParser struct {
	position int
	ch       rune
	input    string
	inputLen int
}

type (
	JysmolValue     interface{}
	JysmolPrimitive interface{}
	JysmolArray     []JysmolValue
	JysmolObject    map[string]JysmolValue
)

func Parse(input string) (JysmolValue, error) {
	p := JysmolParserNew(input)

	p.skipWhitespace()

	return p.parseValue()
}

func JysmolParserNew(input string) *JysmolParser {
	p := &JysmolParser{
		position: -1,
		input:    input,
		inputLen: len(input),
	}

	p.advance()

	return p
}

func (p *JysmolParser) parseValue() (JysmolValue, error) {
	p.skipWhitespace()

	switch p.ch {
	case '"':
		return p.parseString()
	case '[':
		return p.parseArray()
	case '{':
		return p.parseObject()
	}

	if isAlpha(p.ch) {
		return p.parseKeyword()
	}

	if isDigit(p.ch) {
		return p.parseNumber()
	}

	if p.ch == '-' {
		p.advance()
		if isDigit(p.ch) {
			float, err := p.parseNumber()
			if err != nil {
				return 0, err
			}
			return -(float), nil
		}
	}

	return nil, fmt.Errorf("unexpected token: '%s' at position %d", string(p.ch), p.position)
}

func (p *JysmolParser) parseArray() (JysmolArray, error) {
	var arr JysmolArray
	p.advance()
	p.skipWhitespace()

	for p.ch != ']' && p.position < p.inputLen {
		value, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		arr = append(arr, value)
		p.skipWhitespace()
		if err := p.eat(','); err != nil {
			return nil, err
		}
		p.skipWhitespace()
	}

	if err := p.eat(']'); err != nil {
		return nil, err
	}

	return arr, nil
}

func (p *JysmolParser) parseObject() (JysmolObject, error) {
	obj := JysmolObject{}
	p.advance()
	p.skipWhitespace()

	for p.ch != '}' && p.position < p.inputLen {
		key, err := p.parseString()
		if err != nil {
			return nil, err
		}
		p.skipWhitespace()
		p.eat(':')
		value, err := p.parseValue()
		if err != nil {
			return nil, err
		}
		if err := p.eat(','); err != nil {
			return nil, err
		}
		p.skipWhitespace()
		obj[key] = value
	}

	if err := p.eat('}'); err != nil {
		return nil, err
	}

	return obj, nil
}

func (p *JysmolParser) parseNumber() (float64, error) {
	lit := ""
	l, err := p.parseIntLiteral()
	if err != nil {
		return 0, err
	}
	lit += l

	if p.ch == '.' {
		if err := p.eat('.'); err != nil {
			return 0, err
		}
		lit += "."
		l, err := p.parseIntLiteral()
		if err != nil {
			return 0, err
		}
		lit += l
	}

	return strconv.ParseFloat(lit, 64)
}

func (p *JysmolParser) parseIntLiteral() (string, error) {
	lit := string(p.ch)
	p.advance()

	for isDigit(p.ch) && p.position < p.inputLen {
		lit += string(p.ch)
		p.advance()
	}

	return lit, nil
}

func (p *JysmolParser) parseKeyword() (JysmolPrimitive, error) {
	lit := string(p.ch)
	p.advance()

	for isAlpha(p.ch) && p.position < p.inputLen {
		lit += string(p.ch)
		p.advance()
	}

	switch lit {
	case "null":
		return nil, nil
	case "true":
		return true, nil
	case "false":
		return false, nil
	}

	return nil, fmt.Errorf("invalid keyword: %s", lit)
}

func (p *JysmolParser) parseString() (string, error) {
	lit := ""
	p.advance()

	for p.ch != '"' && p.position < p.inputLen {
		lit += string(p.ch)
		p.advance()
	}

	if err := p.eat('"'); err != nil {
		return "", err
	}

	return lit, nil
}

func (p *JysmolParser) advance() {
	p.position++
	if p.position < p.inputLen {
		p.ch = rune(p.input[p.position])
	} else {
		p.ch = '\000'
	}
}

func (p *JysmolParser) skipWhitespace() {
	for p.ch == ' ' || p.ch == '\n' || p.ch == '\r' || p.ch == '\t' {
		p.advance()
	}
}

func (p *JysmolParser) eat(ch rune) error {
	if p.ch != ch {
		return fmt.Errorf("expected: '%s', got: '%s' at position %d", string(ch), string(p.ch), p.position)
	}
	p.advance()

	return nil
}

func isAlpha(ch rune) bool {
	return ch >= 'a' && ch <= 'z' || ch >= 'A' && ch <= 'Z'
}

func isDigit(ch rune) bool {
	return ch >= '0' && ch <= '9'
}
